File: apps/server/src/cocktail_server/schemas/events.py
from __future__ import annotations

from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from cocktail_server.schemas.messages import Message


class ConversationEvent(BaseModel):
    """新規会話が生成されたことを通知する（既存 `conversation_id` 指定時は飛ばない）。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["conversation"] = "conversation"
    conversation_id: str


class UserSavedEvent(BaseModel):
    """ユーザメッセージが保存された。UI は楽観的表示をサーバ側 id に置き換える。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["user_saved"] = "user_saved"
    message: Message


class AssistantStartEvent(BaseModel):
    """アシスタントのターン開始。以降の token/tool 系イベントはこの `message_id` に紐づく。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["assistant_start"] = "assistant_start"
    message_id: str


class ToolCallStartEvent(BaseModel):
    """ツール呼び出しの開始。UI はスピナーと args を即時表示する。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["tool_call_start"] = "tool_call_start"
    call_id: str
    name: str
    args: dict[str, Any]


class ToolCallEndEvent(BaseModel):
    """ツール呼び出しの終了。status='error' のときは `data.error` にメッセージを入れる。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["tool_call_end"] = "tool_call_end"
    call_id: str
    status: Literal["done", "error"]
    summary: str
    data: dict[str, Any]


class ImageReadyEvent(BaseModel):
    """画像がディスクに書かれて配信可能になった。便宜上 tool_call_end より前に飛ぶ。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["image_ready"] = "image_ready"
    call_id: str
    image_id: str
    image_url: str
    mime: str
    width: int
    height: int


class AssistantEndEvent(BaseModel):
    """アシスタントメッセージが完成。これが UI にとっての正規状態。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["assistant_end"] = "assistant_end"
    message: Message


class ErrorEvent(BaseModel):
    """致命的エラー。ストリームはこのあと `done` で閉じられる。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["error"] = "error"
    code: str
    message: str


class DoneEvent(BaseModel):
    """ストリーム完了。以降イベントは来ない。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    type: Literal["done"] = "done"


SseEvent = Annotated[
    ConversationEvent
    | UserSavedEvent
    | AssistantStartEvent
    | ToolCallStartEvent
    | ToolCallEndEvent
    | ImageReadyEvent
    | AssistantEndEvent
    | ErrorEvent
    | DoneEvent,
    Field(discriminator="type"),
]
"""`POST /chat` が SSE で流す全イベント。"""

File: packages/api-types/src/events.ts
import { z } from "zod";

import { MessageSchema } from "./messages.js";

/**
 * `POST /chat` が SSE で流すイベント。サーバ側 `schemas/events.py` と完全対応。
 * `type` フィールドは SSE の `event:` 名とも一致するので、ストリーム側でも
 * 同じ string で分岐してよい。
 */

export const ConversationEventSchema = z
  .object({
    type: z.literal("conversation"),
    conversation_id: z.string(),
  })
  .strict();

export const UserSavedEventSchema = z
  .object({
    type: z.literal("user_saved"),
    message: MessageSchema,
  })
  .strict();

export const AssistantStartEventSchema = z
  .object({
    type: z.literal("assistant_start"),
    message_id: z.string(),
  })
  .strict();

export const ToolCallStartEventSchema = z
  .object({
    type: z.literal("tool_call_start"),
    call_id: z.string(),
    name: z.string(),
    args: z.record(z.unknown()),
  })
  .strict();

export const ToolCallEndEventSchema = z
  .object({
    type: z.literal("tool_call_end"),
    call_id: z.string(),
    status: z.enum(["done", "error"]),
    summary: z.string(),
    data: z.record(z.unknown()),
  })
  .strict();

export const ImageReadyEventSchema = z
  .object({
    type: z.literal("image_ready"),
    call_id: z.string(),
    image_id: z.string(),
    image_url: z.string(),
    mime: z.string(),
    width: z.number().int(),
    height: z.number().int(),
  })
  .strict();

export const AssistantEndEventSchema = z
  .object({
    type: z.literal("assistant_end"),
    message: MessageSchema,
  })
  .strict();

export const ErrorEventSchema = z
  .object({
    type: z.literal("error"),
    code: z.string(),
    message: z.string(),
  })
  .strict();

export const DoneEventSchema = z
  .object({
    type: z.literal("done"),
  })
  .strict();

export const SseEventSchema = z.discriminatedUnion("type", [
  ConversationEventSchema,
  UserSavedEventSchema,
  AssistantStartEventSchema,
  ToolCallStartEventSchema,
  ToolCallEndEventSchema,
  ImageReadyEventSchema,
  AssistantEndEventSchema,
  ErrorEventSchema,
  DoneEventSchema,
]);
export type SseEvent = z.infer<typeof SseEventSchema>;

export type ConversationEvent = z.infer<typeof ConversationEventSchema>;
export type UserSavedEvent = z.infer<typeof UserSavedEventSchema>;
export type AssistantStartEvent = z.infer<typeof AssistantStartEventSchema>;
export type ToolCallStartEvent = z.infer<typeof ToolCallStartEventSchema>;
export type ToolCallEndEvent = z.infer<typeof ToolCallEndEventSchema>;
export type ImageReadyEvent = z.infer<typeof ImageReadyEventSchema>;
export type AssistantEndEvent = z.infer<typeof AssistantEndEventSchema>;
export type ErrorEvent = z.infer<typeof ErrorEventSchema>;
export type DoneEvent = z.infer<typeof DoneEventSchema>;
